fastscanner/benchmarks/benchmark_redis_read.py

```diff
@@ -0,0 +1,127 @@
+import asyncio
+import json
+import logging
+import os
+import time
+from datetime import datetime
+
+import redis.asyncio as aioredis
+
+from fastscanner.pkg import config
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+SYMBOLS_FILE = "data/symbols/symbols.json"
+STREAM_PREFIX = "candles_min_"
+NO_DATA_TIMEOUT = 10
+
+batch_start_time = None
+last_received_time = None
+batch_lock = asyncio.Lock()
+total_messages = 0
+
+
+async def get_or_load_symbols(redis: aioredis.Redis) -> list[str]:
+    os.makedirs(os.path.dirname(SYMBOLS_FILE), exist_ok=True)
+
+    if os.path.exists(SYMBOLS_FILE):
+        with open(SYMBOLS_FILE, "r") as f:
+            symbols = json.load(f)
+            logger.info(f"Loaded {len(symbols)} symbols from file.")
+            return symbols
+
+    symbols = []
+    async for key in redis.scan_iter(f"{STREAM_PREFIX}*"):
+        if await redis.type(key) == "stream":
+            symbol = key.replace(STREAM_PREFIX, "")
+            symbols.append(symbol)
+
+    with open(SYMBOLS_FILE, "w") as f:
+        json.dump(symbols, f)
+
+    logger.info(f"Discovered and saved {len(symbols)} symbols.")
+    return symbols
+
+
+async def read_stream(redis: aioredis.Redis, symbol: str):
+    global batch_start_time, last_received_time, total_messages
+
+    stream_key = f"{STREAM_PREFIX}{symbol}"
+    last_id = "$"
+
+    logger.info(f"Listening for stream: {stream_key}")
+
+    try:
+        while True:
+            entries = await redis.xread({stream_key: last_id}, block=1000, count=100)
+            now = time.time()
+            if entries:
+                for _, stream_entries in entries:
+                    for entry_id, data in stream_entries:
+                        ts = float(data.get("timestamp", now))
+                        logger.info(
+                            f"[{symbol}] ID: {entry_id}, Timestamp: {ts}, Data: {data}"
+                        )
+                        last_id = entry_id
+
+                        async with batch_lock:
+                            if batch_start_time is None:
+                                batch_start_time = now
+                            last_received_time = now
+                            total_messages += 1
```

Comment:
```suggestion
              for _, stream_entries in entries:
                  for entry_id, data in stream_entries:
                      ts = float(data.get("timestamp", now))
                      logger.info(
                          f"[{symbol}] ID: {entry_id}, Timestamp: {ts}, Data: {data}"
                      )
                      last_id = entry_id
    
                      if batch_start_time is None:
                          batch_start_time = now
                      last_received_time = now
                      total_messages += 1
```
Be careful with deep nesting. It reduces readability. The code starts looking like it was frontend 😅 
The `if entries` is not necessary because if the dict is empty, it won't run any iteration.
No need to lock there as that part of the code is synchronous.